src/main/java/org/example/MediumAI.java

```diff
@@ -0,0 +1,53 @@
+package org.example;
+
+import static tictactoe.Main.*;
+
+public class MediumAI extends Player{
+    @Override
+    public char[][] play(char[][] matrix, char yourChar) {
+        if (checkWinner(matrix).equals("Game not finished")) {
+            System.out.println("Making move level \"medium\"");
+            //1st rule
+            //Instant win
+            char[][] instantWin = instantWin(matrix, yourChar);
+            if(instantWin != null) return instantWin;
+            //2nd rule
+            //Prevent loss
+            char[][] preventLoss = preventLoss(matrix, yourChar);
+            if(preventLoss != null) return preventLoss;
+            //3rd rule
+            //Do random move
+            matrix = EasyAI.randomMoveGenerator(matrix, yourChar);
+        }
+        return matrix;
+    }
+
+    static char[][] instantWin(char[][] matrix, char yourChar) {
+        for (int i = 0; i < matrix.length; i++) {
+            for (int j = 0; j < matrix[i].length; j++) {
+                if (matrix[i][j] == ' ') {
+                    matrix[i][j] = otherChar(yourChar);
+                    if (checkCharWinner(matrix, otherChar(yourChar))) {
+                        matrix[i][j] = yourChar;
+                        return matrix;
+                    } else matrix[i][j] = ' ';
+                }
+            }
+        }
+        return null;
```

Comment:
But in the play() method of this class I rely on the part if the result of this method is null or not. If it is null, this means matrix is not changed when checking for instant wins or preventing instant loss. The play method relies on this logic: 1. if there's instant win, do it and return, if not - 2. check if there's a way to prevent loss, if yes do that and return if not - 3. Make a random move. I return null, to go on with this logic. Do you think it's still better here not to return null?